Rework Motivation; other smaller edits (#2)
kavon/swift-evolution: proposals/NNNN-noncopyable-generics.md

This ensures correct semantics for types for which it is not meaningful to
have multiple copies:
```swift
// A file descriptor cannot be usefully used from
// multiple places at once, so make it non-copyable
// to prevent such usage.
struct FileDescriptor: ~Copyable { ... }
```

This also provides an alternative to class objects for some use cases.
In particular, the tightly defined lifecycle allows noncopyable structs
to carry deinitializers:
```swift
struct HeapBuffer: ~Copyable {
  init() { ... allocate working storage on heap ... }
  deinit() { ... release storage ... }
}
```

But SE-0390 also made a number of concessions to simplify the initial implementation,
limitations which significantly reduce the usefulness of noncopyable types.

For example, if noncopyable types cannot be used in generics,
then they cannot be used with `Optional`,
which prevents you from defining failable initializers:
```swift
struct FileDescriptor: ~Copyable {
  init?(filename: String) { // 🛑 Cannot return Optional<FileDescriptor>
    ...
  }
}
```

Practical use of generics also requires supporting protocol conformances,

> since generic parameters are often constrained to particular protocols:

Reviewer: ```suggestion
Practical use of generics also requires supporting protocol conformances,
since generic parameters gain capabilities when they are constrained to particular protocols:
```

```swift

> func max<T: Comparable>(...) { ... }

Reviewer: This explanation subtly reminds readers what "positive constraints" really mean in the language: they announce what a generic value is capable of doing, not what kind of values they exclude.
```suggestion
// T is capable of being compared with other T's
func max<T: Comparable>(...) { ... }
```

```

In order to broaden the expressiveness and utility of noncopyable types, then,
we need a sound way to allow these types to be used in generic parameters,

> to conform to protocols, and to be stored in existentials.

Reviewer: ```suggestion
In order to broaden the expressiveness and utility of noncopyable types, then,
language extensions are needed to allow these types to be used in generic parameters,
to conform to protocols, and to be stored in existentials.
```

2. This protocol is applied by default to type definitions and generic requirements
3. The `~Copyable` notation can suppress this implicit requirement in specific cases

> **Note**: Several other issues will need to be addressed before we can adapt the bulk of the standard library to support noncopyable types.
> We are exploring possible approaches and hope to have a concrete proposal in the near future.
> This expansion of the generics system is an obvious prerequisite for any such effort.

Reviewer: ```suggestion
**Note**: Several other issues will need to be addressed before the bulk of the standard library can be adapted to support noncopyable types.
Various approaches are being explored for separate proposals in the near future.
This proposal's expansion of the generics system is a clear prerequisite for any such effort.
```
